### PR TITLE
Rework to support airgapped cluster

### PR DESCRIPTION
**Pull request by nolancon**
This is a bit of a monster so I'll summarise the main changes
 - The `uninstall` command now accepts flags for yamls such as `--resource-quota.yaml` etc that are already in place for the `install` command. This is so that the plugin willl not attempt to pull these yaml files from dockerhub in an airgapped environment.
 - The `NewInstaller` function was getting hard to manage as it was trying to satisfy all commands. So now we have separate funcs eg `NewInstaller` for `install` command, `NewUninstaller` for` uninstall` command and `NewPortalManagerInstaller` for portal related commands (`install-portal`, `unisntall-portal`, `disable-portal` and `enable-portal`).
 - This allows us to use a new struct (`installerOptions`) to **only** fetch the manifests we require for a given command. Where previously we were building an entire manifest in-memory filesystem unnecessarily, now we only use the manifests that will be required by the command. This means that airgapped installations are much less complex.
 
 Tested the following "happy path" `install` and `uninstall` commands on an an airgapped cluster and all seems to be in order 🤞 
 
```
 ./kubectl-storageos install   \
 --stos-cluster-yaml ondat.yaml \
 --etcd-endpoints storageos-etcd-client.storageos-etcd:2379 \
 --stos-operator-yaml operator.yaml \
 --stos-version v2.5.0
```
```
 ./kubectl-storageos uninstall  \
 --stos-cluster-yaml ondat.yaml \
 --stos-operator-yaml operator.yaml
```

*Note*: `upgrade` command is not supported in an airgapped cluster yet

### CODE REVIEW

File: pkg/consts/consts.go [hunk 1]
```diff
@@ -2,6 +2,7 @@ package consts
 
 const (
 	OperatorOldestSupportedVersion = "v2.2.0"
+	ClusterOperatorLastVersion     = "v2.4.4"
```
Comment (mhmxs) [inline pkg/consts/consts.go:5]: m i right this means they are only able to use the function if they have upgraded the old cluster to the latest?

File: pkg/installer/filebuilder.go [hunk 2]
```diff
@@ -235,3 +258,11 @@ func (fb *fileBuilder) readOrPullManifest(config *rest.Config) (string, error) {
 
 	return string(contents), nil
 }
+
+// getYamlPath returns whichever yamlPath is set (no more than one will ever be set)
+func getYamlPath(installYamlPath, uninstallYamlPath string) string {
```
Comment (mhmxs) [inline pkg/installer/filebuilder.go:263]: This should be a more generic function, like `getStringWithDefault` vale.

File: pkg/installer/installer.go [hunk 3]
```diff
@@ -244,24 +299,97 @@ func NewDryRunInstaller(config *apiv1.KubectlStorageOSConfig) (*Installer, error
...
+	}
+
+	distribution := pluginutils.DetermineDistribution(currentVersion.String())
+
+	kubesystemNS, err := pluginutils.GetNamespace(clientConfig, "kube-system")
+	if err != nil {
+		return uninstaller, errors.WithStack(err)
+	}
+
+	uninstallPortal := false
```
Comment (mhmxs) [inline pkg/installer/installer.go:352]: This should be initialized just before `if stosCluster != nil {` where it is actually used

File: pkg/version/version.go [hunk 4]
```diff
@@ -138,6 +138,30 @@ func GetExistingOperatorVersion(namespace string) (string, error) {
 	return version, nil
 }
 
+func GetExistingEtcdOperatorVersion(namespace string) (string, error) {
+	if namespace == "" {
+		namespace = "storageos-etcd"
```
Comment (mhmxs) [inline pkg/version/version.go:143]: WE have a constent for this: `EtcdOperatorNamespace`

File: pkg/version/version.go [hunk 5]
```diff
@@ -138,6 +138,30 @@ func GetExistingOperatorVersion(namespace string) (string, error) {
...
+	config, err := pluginutils.NewClientConfig()
+	if err != nil {
+		return "", err
+	}
+
+	clientset, err := pluginutils.GetClientsetFromConfig(config)
+	if err != nil {
+		return "", errors.Wrap(err, consts.ErrUnableToContructClientFromConfig)
+	}
+	etcdDeployment, err := clientset.AppsV1().Deployments(namespace).Get(context.TODO(), "storageos-etcd-controller-manager", metav1.GetOptions{})
```
Comment (mhmxs) [inline pkg/version/version.go:154]: `storageos-etcd-controller-manager` may need a constant, because it is used in multiple places

**Review by mhmxs — COMMENTED** [review]
I love this change

File: pkg/consts/consts.go (hunk 1 above)
Comment (nolancon) [reply pkg/consts/consts.go:5]: I'm not sure I understand the question, but to clarify: 
We had a function to dynamically discover the old cluster-operator "last version" (i.e the latest release). But now that cluster-operator is no longer active, the "last version" is static at v2.4.4. 

This version is only used to check if the installed operator (discovered via image tag) is lower than this. If it is, we use that installed operator version to then pull the manifests for the uninstall process.

**Review by mhmxs — APPROVED** [review]

**Pull request merged**